File: middle_of_the_linked_list876.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def middleNode(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        l = r = head
        while r.next is not None:
            l = l.next
            r = r.next
            if r.next is None:
                break
            else:
                r = r.next
        
        return l

File: middle_of_the_linked_list876.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* middleNode(struct ListNode* head) {
    struct ListNode *slow = head, *fast = head;
    while (fast->next != NULL) {
        slow = slow->next;
        fast = fast->next;
        if (fast->next == NULL) {
            break;
        } else {
            fast = fast->next;
        }
    }
    return slow;
}